File: src/context.ts
import {
  Temporal,
  systemNow,
  systemTimeZone,
  toZoned,
  type Instant,
  type Zoned,
} from './clock/index.js';
import type { DayPeriodRule } from './data/dayPeriods.js';
import { firstDayForRegion } from './data/weekData.js';
import type { Weekday } from './ir/types.js';

export class ConfigError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ConfigError';
  }
}

export type DateOrder = 'MDY' | 'DMY' | 'YMD';
export type Bias = 'past' | 'future' | 'none';
/** The dialectal "next Tuesday" question: nearest upcoming vs Tuesday of next week. */
export type NextWeekdayPolicy = 'nearest' | 'week-after';
/** Does "the last 3 days" include the partial current day? */
export type PartialPeriodPolicy = 'include' | 'exclude';

export interface TimeContextOptions {
  /**
   * Reference instant. Defaults to the system clock; pass an explicit value
   * (Temporal.Instant or ISO string) for reproducible resolution and tests.
   */
  now?: Instant | string;
  /** IANA timezone of the *user/utterance*, not the machine. Defaults to system. */
  timeZone?: string;
  /** BCP-47 language-REGION tag. Region drives week-start & date-order defaults. */
  locale?: string;
  weekStart?: Weekday;
  dateOrder?: DateOrder;
  /** Direction preference for underspecified expressions ("Friday", "May 29"). */
  bias?: Bias;
  nextWeekday?: NextWeekdayPolicy;
  partialPeriod?: PartialPeriodPolicy;
  /** Override the locale's day-period boundaries (e.g. to match another system's conventions). */
  dayPeriods?: DayPeriodRule[];
}

const MDY_REGIONS = ['US', 'PH', 'UM', 'VI', 'GU', 'AS', 'PR'];
const YMD_REGIONS = ['CN', 'JP', 'KR', 'TW', 'HU', 'MN', 'LT'];

export class TimeContext {
  readonly now: Instant;
  readonly timeZone: string;
  readonly locale: string;
  readonly language: string;
  readonly region: string | undefined;
  readonly weekStart: Weekday;
  readonly dateOrder: DateOrder;
  readonly bias: Bias;
  readonly nextWeekday: NextWeekdayPolicy;
  readonly partialPeriod: PartialPeriodPolicy;
  readonly dayPeriods: DayPeriodRule[] | undefined;

  private constructor(opts: TimeContextOptions) {
    const rawNow = opts.now ?? systemNow();
    try {
      this.now = typeof rawNow === 'string' ? parseInstant(rawNow) : rawNow;
    } catch (e) {
      throw new ConfigError(`invalid now: ${String(e)}`);
    }
    this.timeZone = opts.timeZone ?? systemTimeZone();
    this.locale = opts.locale ?? 'en-US';

    let language: string;
    let region: string | undefined;
    try {
      const tag = new Intl.Locale(this.locale);
      language = tag.language;
      region = tag.region;
    } catch {
      throw new ConfigError(`invalid locale: ${this.locale}`);
    }
    this.language = language;
    this.region = region;

    this.weekStart = opts.weekStart ?? firstDayForRegion(region);
    this.dateOrder =
      opts.dateOrder ??
      (region && MDY_REGIONS.includes(region)
        ? 'MDY'
        : region && YMD_REGIONS.includes(region)
          ? 'YMD'
          : 'DMY');
    this.bias = opts.bias ?? 'none';
    this.nextWeekday = opts.nextWeekday ?? 'nearest';
    this.partialPeriod = opts.partialPeriod ?? 'include';
    this.dayPeriods = opts.dayPeriods;

    // Validate the timezone eagerly so failures surface at construction.
    try {
      toZoned(this.now, this.timeZone);
    } catch {
      throw new ConfigError(`invalid timeZone: ${this.timeZone}`);
    }
  }

  static make(opts: TimeContextOptions = {}): TimeContext {
    return new TimeContext(opts);
  }

  /** The reference instant projected into the context timezone. */
  get zonedNow(): Zoned {
    return toZoned(this.now, this.timeZone);
  }
}

function parseInstant(s: string): Instant {
  return Temporal.Instant.from(s);
}

File: python/nl2time/context.py
"""TimeContext: all resolution policy flows through here (port of src/context.ts).

Options use the same camelCase names as the JS reference and the corpus wire
format (`timeZone`, `weekStart`, ...); attributes are snake_case.
"""

from __future__ import annotations

from typing import Any

from whenever import Instant, ZonedDateTime

from .clock import parse_instant, system_now, system_time_zone, to_zoned
from .data import DayPeriodRule, first_day_for_region
from .errors import ConfigError

_MDY_REGIONS = frozenset(["US", "PH", "UM", "VI", "GU", "AS", "PR"])
_YMD_REGIONS = frozenset(["CN", "JP", "KR", "TW", "HU", "MN", "LT"])

_WEEKDAYS = frozenset(["mon", "tue", "wed", "thu", "fri", "sat", "sun"])


def _parse_locale(locale: str) -> tuple[str, str | None]:
    """Split a BCP-47 tag without Intl: language = first subtag; region = the
    first 2-letter-alpha or 3-digit subtag after it (script subtags skipped)."""
    parts = locale.replace("_", "-").split("-")
    language = parts[0].lower()
    if not language or not language.isascii() or not language.isalpha():
        raise ConfigError(f"invalid locale: {locale}")
    region: str | None = None
    for sub in parts[1:]:
        if (len(sub) == 2 and sub.isascii() and sub.isalpha()) or (
            len(sub) == 3 and sub.isdigit()
        ):
            region = sub.upper()
            break
    return language, region


def _normalize_day_periods(rules: list[dict[str, Any]]) -> list[DayPeriodRule]:
    """Accept wire-format rules ({period, from, before}) or internal from_."""
    out: list[DayPeriodRule] = []
    for r in rules:
        out.append(
            {
                "period": r["period"],
                "from_": r["from_"] if "from_" in r else r["from"],
                "before": r["before"],
            }
        )
    return out


class TimeContext:
    def __init__(
        self,
        *,
        now: Instant | str | None = None,
        timeZone: str | None = None,
        locale: str | None = None,
        weekStart: str | None = None,
        dateOrder: str | None = None,
        bias: str | None = None,
        nextWeekday: str | None = None,
        partialPeriod: str | None = None,
        dayPeriods: list[dict[str, Any]] | None = None,
    ) -> None:
        raw_now = now if now is not None else system_now()
        try:
            self.now: Instant = (
                parse_instant(raw_now) if isinstance(raw_now, str) else raw_now
            )
        except Exception as e:  # noqa: BLE001 - mirror JS ConfigError wrapping
            raise ConfigError(f"invalid now: {e}") from e
        self.time_zone: str = timeZone if timeZone is not None else system_time_zone()
        self.locale: str = locale if locale is not None else "en-US"

        self.language, self.region = _parse_locale(self.locale)

        if weekStart is not None and weekStart not in _WEEKDAYS:
            raise ConfigError(f"invalid weekStart: {weekStart}")
        self.week_start: str = (
            weekStart if weekStart is not None else first_day_for_region(self.region)
        )
        if dateOrder is not None:
            self.date_order = dateOrder
        elif self.region in _MDY_REGIONS:
            self.date_order = "MDY"
        elif self.region in _YMD_REGIONS:
            self.date_order = "YMD"
        else:
            self.date_order = "DMY"
        self.bias: str = bias if bias is not None else "none"
        self.next_weekday: str = nextWeekday if nextWeekday is not None else "nearest"
        self.partial_period: str = (
            partialPeriod if partialPeriod is not None else "include"
        )
        self.day_periods: list[DayPeriodRule] | None = (
            _normalize_day_periods(dayPeriods) if dayPeriods is not None else None
        )

        # Validate the timezone eagerly so failures surface at construction.
        try:
            to_zoned(self.now, self.time_zone)
        except Exception as e:  # noqa: BLE001
            raise ConfigError(f"invalid timeZone: {self.time_zone}") from e

    @classmethod
    def make(cls, opts: dict[str, Any] | None = None, /, **kwargs: Any) -> "TimeContext":
        merged = {**(opts or {}), **kwargs}
        return cls(**merged)

    @property
    def zoned_now(self) -> ZonedDateTime:
        """The reference instant projected into the context timezone."""
        return to_zoned(self.now, self.time_zone)
